### PR TITLE
[DATA-283] Implement client.py with Global Client Auth

### PR DESCRIPTION
**Pull request by abhinavpottabathula**
[Jira Ticket](https://runelabs.atlassian.net/browse/DATA-283)

[Design Doc](https://docs.google.com/document/d/1QDfQvEx12iJBiYz4sTa1dO4RCBfsKBgfLrUnYmKR8LA/edit#heading=h.a29kgy9j22qh)

Notes for Reviewers:
- I based `runeq/v2sdk/__init__.py` on the [datadog sdk](https://github.com/DataDog/datadogpy/blob/master/datadog/__init__.py#L33-L137) which uses global initialization for credentials.
- Added `get_stream` and `get_availability` queries  in `stream.py` in order to test that initializing credentials with `initialization()` actually works.
- I didn't add any graph client tests, because the stream client tests cover all possible ways to authenticate a user with `initialize()`. If adding in graph client tests is useful, then I'd need to define a minimum query (maybe whoami?) that would be useful to add and test. However, I think this more relevant for future PRs with SDK support for graphql queries.
- Should there be a `404 NotFound` error when a user with invalid credentials queries for a non existent stream?
- I don't particularly like how `get_stream` and `get_availability` peak into the stream client: `client._stream_client.config.stream_url`, but I couldn't think of a cleaner way of doing it.
- It would be nice to have a type def for `_gql_client` and `_stream_client` but the type def can't exist before the class def. Does anyone know a clean work around?

### CODE REVIEW

File: runeq/v2sdk/stream.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+"""
+V2 SDK functionality to support Stream operations.
+
+"""
+from runeq import errors
+from . import client
+
+
+def get_stream(stream_id, **params):
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:9]: This could use some more documentation about how its return type is intended to be used---looking at how `_get` works, it seems that it's returning a generator, but that's not obvious unless you poke around a few levels down.

File: runeq/v2sdk/stream.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            limit: Limits the number of points returned for the stream.
+            format: 'csv' or 'json'. Defaults to csv.
+    """
+    if not client._stream_client:
+        raise errors.InitializationError("Failed to initialize stream client.")
+
+    stream_url = client._stream_client._config.stream_url
+    url = f"{stream_url}/v2/streams/{stream_id}/availability"
+
+    # TODO: Currently it seems that only JSON availability is supported, but
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:44]: CSV should be supported (the same as the streams endpoint: with CSV by default and JSON if you pass a `format=json` param in the querystring). The linked ticket is about supporting batch availability (where you can ask for the availability for multiple streams at a time).

File: runeq/v2sdk/__init__.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            "WARNING: You are accessing a part of the Rune SDK that is "
+            "currently in Beta. Please be aware that Beta functionality is "
+            "subject to change in future releases, and may break your code "
+            "upon upgrade.",
+            file=stderr,
+        )
+        _beta_warned = True
+
+
+def initialize(*args, **kwargs):
```
Comment (maxmora) [inline runeq/v2sdk/__init__.py:36]: This is a pretty high-level question, but I don't think I see an answer in the design doc: Is there a reason that the SDK is based around global state and an `initialize` function, rather than being a bit more object oriented (where you instantiate a client and call methods on that)?

**Comment by maxmora** [comment]
> * Should there be a `404 NotFound` error when a user with invalid credentials queries for a non existent stream?

Is this a question about the what the SDK should do, or about what the API does? It looks like the API does 404 if a stream doesn't exist, regardless of the validity of the auth credentials, though maybe that's something we should change.

> * It would be nice to have a type def for `_gql_client` and `_stream_client` but the type def can't exist before the class def. Does anyone know a clean work around?

I'd be inclined to reorganize the code and put the singleton `_gql_client` and `_stream_client` variables where they're used (in `runeq/v2sdk/stream.py` and wherever the wrapper functions around the functions will live), rather than next to where their classes are defined. I think that should work without any circular import errors.

File: runeq/v2sdk/__init__.py (hunk 3 above)
Comment (abhinavpottabathula) [reply runeq/v2sdk/__init__.py:36]: @cranti suggested this idea in [this comment](https://docs.google.com/document/d/1D2WFmc4OgSssKe7Ae9FrvpoTItAS_k6245Yk2UZ_93M/edit?disco=AAAAcWo1cvk) on an older design doc.

File: runeq/v2sdk/client.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+import requests
+
+from gql import Client as GQLClient
+from gql.transport.requests import RequestsHTTPTransport
+
+from runeq import Config, errors
+
+
+_gql_client = None
```
[the remark this answers is not in the document]
Comment (abhinavpottabathula) [reply runeq/v2sdk/client.py:14]: This conversation is being continued in the comment thread about reorganizing the code and putting the singleton _gql_client and _stream_client variables where they're used.

**Comment by abhinavpottabathula** [comment]
> > * Should there be a `404 NotFound` error when a user with invalid credentials queries for a non existent stream?
> 
> Is this a question about the what the SDK should do, or about what the API does? It looks like the API does 404 if a stream doesn't exist, regardless of the validity of the auth credentials, though maybe that's something we should change.
>

I was thinking that the API should be changed to prioritize consistency between queries with access keys and client keys. Whether both `test_init_with_access_keys` and `test_init_with_client_keys` return `401` or `404` isn't as important, but I think the API should return `401` in the case of invalid auth, and then `404` with valid auth and invalid stream id. Just so that no extra information is being leaked, namely a bad actor could provide incorrect credentials and find valid stream ids.
> > * It would be nice to have a type def for `_gql_client` and `_stream_client` but the type def can't exist before the class def. Does anyone know a clean work around?
> 
> I'd be inclined to reorganize the code and put the singleton `_gql_client` and `_stream_client` variables where they're used (in `runeq/v2sdk/stream.py` and wherever the wrapper functions around the functions will live), rather than next to where their classes are defined. I think that should work without any circular import errors.

The singleton `_gql_client` and `_stream_client` variables are intended to be used in several files (patient.py, devices.py, etc...) in order to query for data. It isn't obvious to me which of those files should have the singleton variables, so I decided to put it in where the clients are defined.

File: runeq/v2sdk/stream.py [hunk 5]
```diff
@@ -6,42 +6,55 @@
 from . import client
 
 
-def get_stream(stream_id, **params):
+def get_stream(stream_id: str, **params):
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:9]: It seems to me it would be preferable to have separate parameters (with type annotations) so the SDK supports static type analysis (and is a bit more self-documenting). This should also help IDEs and such spot errors early (which would be help the UX I imagine, since this library is intended for non-developers as well). The code below can combine those into a dict for the purposes of making the GET request. This method could also have a return type.

This comment applies to `get_availability` below as well.

File: runeq/v2sdk/stream.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    stream_id : str
+        Unique string identifier for a stream.
+
+    **params
+        limit: Limits the number of points returned for the stream.
+        format: 'csv' or 'json'. Defaults to csv.
+
+    Returns
+    -------
+    Generator over stream data in the specified format, defaulting to csv.
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:24]: I think this could be documented a bit more clearly (with the same comment for the availability method): With CSV, each iteration will return a string for the response worth of CSV, whereas with JSON it comes back as a dictionary containing the the response body. In other words, the default is to returned unstructured data that would require further parsing to use, whereas specifying JSON returns structured data (though you still need to read the API docs to know what that structure is).

File: runeq/v2sdk/stream.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    Fetch stream data from the Stream API.
+
+    Parameters
+    ----------
+    stream_id : str
+        Unique string identifier for a stream.
+
+    **params
+        limit: Limits the number of points returned for the stream.
+        format: 'csv' or 'json'. Defaults to csv.
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:20]: Maybe this has already been discussed, but as a UX consideration, I do wonder if `csv` is really the reasonable default for the SDK to use, but maybe it's common to pass string blobs containing CSV data to analysis libraries?

File: runeq/v2sdk/stream.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    Returns
+    -------
+    Generator over stream data in the specified format, defaulting to csv.
+
+    """
+    if not client._stream_client:
+        raise errors.InitializationError("Failed to initialize stream client.")
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:28]: I feel like the following could be a more useful message, especially for the target audience of these helper functions (non-developers), something like:
```suggestion
        raise errors.InitializationError("runeq must be initialized by calling `initialize` before this function can be used")
```

Even for a developer, this isn't a failure to initialized, it's just that the client isn't initialized, so at this point it's misleading to say it "failed".

File: runeq/v2sdk/stream.py (hunk 7 above)
Comment (abhinavpottabathula) [reply runeq/v2sdk/stream.py:20]: I picked csv as the default because our internal neuro team transform the data they pull from the V1 API into pandas dataframes immediately before working with it. Pandas has a way to easily convert a csv into a dataframe, so it would be more efficient to store it as a csv.

Comment (maxmora) [reply runeq/v2sdk/stream.py:20]: Ah cool, makes sense!

File: runeq/v2sdk/stream.py (hunk 6 above)
Comment (abhinavpottabathula) [reply runeq/v2sdk/stream.py:24]: Adding better documentation, but I am a little confused what you mean by the CSV response being unstructured data and the JSON response being structured. The csv data should be structured because it contains the column names in every response.

File: runeq/v2sdk/stream.py [hunk 9]
```diff
@@ -6,7 +6,7 @@
 from . import client
 
 
-def get_stream(stream_id: str, **params):
+def get_stream(stream_id: str, limit: int = 100, format: str = "csv"):
```
Comment (maxmora) [inline runeq/v2sdk/stream.py:9]: `format` can be `Literal["csv", "json"]` (from package `typing`) for the type annotations to enforce those two values specifically when doing static analysis.

File: runeq/v2sdk/stream.py (hunk 6 above)
Comment (abhinavpottabathula) [reply runeq/v2sdk/stream.py:27]: Ahh I see what you mean now, made the change!

File: runeq/v2sdk/stream.py (hunk 9 above)
Comment (abhinavpottabathula) [reply runeq/v2sdk/stream.py]: Looks like this isn't supported in python 3.6 and 3.7, which the SDK wants to support.

Comment (abhinavpottabathula) [reply runeq/v2sdk/stream.py]: Opted to raise a value error as an alternative implementation:
```
format_types = ['csv', 'json']
if format not in format_types:
     raise ValueError("Invalid format type. Expected one of: %s" % format_types)
```
